### PR TITLE
Sqrt price x96 helper

### PR DESCRIPTION
**Pull request by RichJamo**
I have added the following code:

1. functions within the UniswapV3Integration contract to get the sqrtPriceX96 from the pool contract for two tokens, to convert from this sqrtPriceX96 to a price ratio, and back again.
2. a new test which tests the approveAndSwap function in UniswapV3Integration, with the following features: a check of the oracle price for the two tokens, a check of the sqrtPriceX96, conversion into a price ratio, and then application of a slippage percentage to this price ratio, before converting back to sqrtPriceX96 and using this as an input parameter to the approveAndSwap function

There is still further work to be done here, but with certain questions to be answered first:

1. The oracle price could also be used as a slippage check, which would guard against front-running attacks in a way that reading sqrtPriceX96 from the pool contract doesn't do - but in order to do this we would possibly need to change the swap functions in the helper to take amountOutMin as an input param
2. The price ratio could be converted to an actual price and then slippage could be applied at that level - this is probably closer to emulating what happens in the Uniswap FE
3. Make the deadline param accessible in the swap helper functions
4. Make the fee param accessible in the swap helper functions

I'm happy to continue this work in these directions, but thought it good to submit this PR at this point just to get feedback as to whether this is moving in the right direction.

### CODE REVIEW

**Review by RichJamo — COMMENTED** [review]
Self reviewed

**Comment by RichJamo** [comment]
I've just noticed that the test succeeds when I run it against the mainnet fork, but fails when run in standard mode (foundry test). I'm busy investigating and setting up mocks so that it can succeed in both.

**Comment by RichJamo** [comment]
> I've just noticed that the test succeeds when I run it against the mainnet fork, but fails when run in standard mode (foundry test). I'm busy investigating and setting up mocks so that it can succeed in both.

Ok, I solved this by rather creating the fork in the test itself. I think it makes more sense this way, and makes more sense to test this on a forked node. However, if you'd prefer to do it another way, let me know👍

**Comment by kopy-kat** [comment]
just did a quick pass - I think overall this looks good! Could you please run `forge fmt` as the formatter to keep the same formatting rules as the remaining code.

When trying to run this on my device, I get a dependency issue with OZ. I think the reason for this is how we manage our pnpm deps (hoisted) and because chainlink requires a different OZ version than eth-infinitism. I think a short term solution would just be to get rid of the chainlink dep and just copy over the interface (think it doesnt have any imports).

File: test/SwapTest.t.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    address internal constant WETH_HOLDER =
+        0x57757E3D981446D585Af0D9Ae4d7DF6D64647806; // account with WETH holdings
+
+    address constant USDC_ADDRESS = 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48;
+    address constant WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    function setUp() public override {
+        string memory MAINNET_RPC_URL = vm.envString("MAINNET_RPC_URL");
+        vm.createSelectFork(MAINNET_RPC_URL);
+
```
Comment (kopy-kat) [inline test/SwapTest.t.sol:33]: I think it'd make sense to use a specific block here - that way the state is cached and tests are fast (except for the first one). Eg ` vm.rollFork(20_426_591);` should work

**Comment by kopy-kat** [comment]
Could you please also run `forge fmt` to make sure the code is formatted properly

**Review by RichJamo — COMMENTED** [review]
Suggested changes made - ran forge fmt, added specific block number for forked test

**Comment by kopy-kat** [comment]
this looks good to me - thank you @RichJamo

**Pull request merged**